src/patient-registration/input/basic-input/input/input.component.tsx

```diff
@@ -30,8 +40,9 @@ export const Input: React.FC<InputProps> = ({ type, label, name, showRequiredAst
           aria-label={name}
           placeholder={placeholder}
           {...field}
-          value={field.value !== null ? field.value : ''}
+          value={field.value !== null ? field.value : defaultValue || ''}
```

Comment:
`field.value ?? defaultValue ?? ''`

Reply:
Thanks. Just learned something new. Now there's a problem. This problem existed even in the code I wrote but but my initial solution had unintended consequences so I reverted to the one above.

`field.value ?? defaultValue ?? ''` doesn't have the behaviour I expect. I'd expect that if `field.value` is a falsy value like `''` or `0`, it should try to use `defaultValue`. When the above code, if I submit a default value for say country as `"Cameroon"`, the above code will ignore this value because of this [patient-registration.component.tsx#L64](https://github.com/openmrs/openmrs-esm-patient-registration/blob/master/src/patient-registration/patient-registration.component.tsx#L64)

So I tried to make the `defaultValue` been checked first before falling to `field.value`. When I do this, the default value is set properly but the form behaves weirdly. All the other fields in the form become disabled. I wasn't really sure why that was happening.

Reply:
An even better solution would be to hook the value of `defaultValue` in `field.value` even before it's used here but I'm not sure if thats possible.

Reply:
So I have seen why this is happening. For some reason setting the `value` attribute on the `<input/>` tag makes the tag `readonly`. Now for fields that I didn't submit a `defaultValue` in the address template configuration, somehow their default value is set to an empty string. So `value={defaultValue ?? field.value ?? ''}` becomes `value=''` for the other fields which I didn't pass in a default value in the configuration and for those that have default values like country, it's been set properly and then the field is made `readonly`.

Reply:
Hm the argument with the falsy value is not really convincing. How do you represent a value of `''` then? Only `undefined` and `null` should be considered, which is exactly what `??` does.

In any case the whole thing sounds fishy. To be precise the `<input />` from React sets *props*, the `value` prop is not the `value` attribute. The `value` attribute refers to the "initial value" of the underlying `input` element. This would be closer to the `defaultValue` prop in React. The `value` prop refers to the `value` property of the `input` element.

If the `value` prop is set then React uses the controlled mode on input elements. In this mode you need to react to changes and update the `value`, otherwise the old `value` remains. If you want to use the managed mode then only apply the `defaultValue`. In this case changes later have no impact on the value - as only the initial value can be set.

Reply:
See https://reactjs.org/docs/forms.html#controlled-components for more details.

Reply:
To fix the `defaultValue` been set to an empty string when no value is configured, I've changed this function from 

```javascript
function getFieldValue(field: string, doc: XMLDocument) {
    let fieldElement = doc.getElementsByName(field)[0];
    return fieldElement ? fieldElement.getAttribute('value') : '';
  }
```

to 

```javascript
function getFieldValue(field: string, doc: XMLDocument) {
    let fieldElement = doc.getElementsByName(field)[0];
    return fieldElement ? fieldElement.getAttribute('value') : null;
  }
```

Now the other fields are untouched and are editable. But for every field that I submit a default value, the value is set and then the field is made readonly. Is that the expected behaviour?

Reply:
> Hm the argument with the falsy value is not really convincing. How do you represent a value of '' then? Only undefined and null should be considered, which is exactly what ?? does.

Yes this is true. In some cases `''` is a valid deafult value. But in cases where a value is coming from the template configuration, that value should have precedence over `''`. I fixed that by putting `defaultValue` first like so 
`value={defaultValue ?? field.value ?? ''}`

Reply:
I would rather do `field.value || defaultValue || ''`. Reason: If we have a defaultValue it would now *always* win - independent if something was set or not. This cannot be the right behavior if you ask me. If what you say is true then `field.value || defaultValue || ''` is what you actually want.

Reply:
But `field.value` is just an empty string set inside the patient registration component for initialization. Not sure how that is supposed to take precedence over the value coming from the backend.

Reply:
I think I'm still lost though. 

> If we have a defaultValue it would now always win - independent if something was set or not.

What do you mean here by if something was set or not? You mean set using `Formik` `initialValues` prop or via the backend configuration?

Reply:
That change fixes the problem though. I see, `field.value` gets updated as I type. Is that what's happening?

Reply:
I think we refer to different things here. I refer to `<input>` and to input's `value` prop. I refer to `field.value` as the current value and to `defaultValue` as some default or initial value.

Now under these assumptions if `defaultValue` is *anything* (i.e., non-falsy) then `defaultValue || field.value || ''` will *always* be `defaultValue` and thus never the current value.

Likewise, if my assumptions are correct, then if we use `field.value || defaultValue || ''` we use the current value exclusively when its non-empty, otherwise we fall back to the initial value. I thought this may be strange as when you start with, e.g., `foo` and you remove the three letters, you'd be back at `foo` (as this is the `defaultValue`). But okay...

Maybe my assumptions here are wrong.

(Just one more remark: If `defaultValue` is really something like a starting value / the value from the backend; why isn't it just set as the initial value of the `field.value`? Normally the `{...field}` does everything in formik and you don't reset the `value` explicitly - as you also don't set the `onChange` etc.)

Reply:
@FlorianRappl this should be fixed now. Thanks for the review :)